### PR TITLE
Update nullability for the value parameter in JsonConverter<T>.Write()

### LINKED ISSUE #56645: Nullability of JsonConverter.Write(T value) not correct
The method `JsonConverter<T>.Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)` has a non-nullable `T value` but `null` can be passed if the converter overrides the `bool HandleNull {get}` method and returns true. That method was added in 5.0 so I believe there was miscommunication around the nullability there. The default is `false` meaning `null` is not passed and the serializer handles it.

### API suggestion:
```diff
namespace System.Text.Json.Serialization
{
    public abstract class JsonConverter<T> : System.Text.Json.Serialization.JsonConverter
    {
-       public abstract void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options);
+       public abstract void Write(Utf8JsonWriter writer, T? value, JsonSerializerOptions options);
    }
}
```

This is a compile-time breaking change for all custom converters (which are common) that are compiled in a nullable context.

Note the corresponding ["T? Read()"](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L143) method already supports nullability.

@stephentoub are we still making nullable changes in 6.0 like this?

**UPDATE**: per discussion, we will unannotate the `value` parameter instead of making it nullable. This may prevent consumers from assuming the converter can always handle `null`. Each derived type should specify the appropriate nullability on the `value` parameter.

### PR DESCRIPTION
**Pull request by steveharter**
Resolves https://github.com/dotnet/runtime/issues/56645

### CODE REVIEW

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 1]
```diff
@@ -818,7 +818,13 @@ public abstract partial class JsonConverter<T> : System.Text.Json.Serialization.
         public virtual bool HandleNull { get { throw null; } }
         public override bool CanConvert(System.Type typeToConvert) { throw null; }
         public abstract T? Read(ref System.Text.Json.Utf8JsonReader reader, System.Type typeToConvert, System.Text.Json.JsonSerializerOptions options);
-        public abstract void Write(System.Text.Json.Utf8JsonWriter writer, T value, System.Text.Json.JsonSerializerOptions options);
+        public abstract void Write(
+            System.Text.Json.Utf8JsonWriter writer,
+#nullable disable
+            T
+#nullable restore
+            value,
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:826]: Nit:
```suggestion
#nullable disable
            T value,
#nullable restore
```

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 2]
```diff
@@ -818,7 +818,13 @@ public abstract partial class JsonConverter<T> : System.Text.Json.Serialization.
...
+           System.Text.Json.JsonSerializerOptions options);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:827]: Nit:
```suggestion
            System.Text.Json.JsonSerializerOptions options);
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 3]
```diff
@@ -583,9 +583,15 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
...
-        /// <param name="value">The value to convert.</param>
+        /// <param name="value">The value to convert. <seealso cref="HandleNull"/> to determine if the converter handles a 'null' reference.</param>
         /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
-        public abstract void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options);
+        public abstract void Write(
+            Utf8JsonWriter writer,
+#nullable disable // T may or may not be nullable depending on the derived type's overload.
+            T
+#nullable restore
+            value,
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:593]: Nit:
```suggestion
#nullable disable // T may or may not be nullable depending on the derived type's override of HandleNull
            T value,
#nullable restore
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 4]
```diff
@@ -139,7 +139,7 @@ internal virtual bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, J
         /// <param name="reader">The <see cref="Utf8JsonReader"/> to read from.</param>
         /// <param name="typeToConvert">The <see cref="Type"/> being converted.</param>
         /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
-        /// <returns>The value that was converted.</returns>
+        /// <returns>The value that was converted. <seealso cref="HandleNull"/> to determine if the converter handles a 'null' JSON value.</returns>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:142]: A custom converter could return `null` regardless of whether it is handling null JSON tokens. I think the `HandleNull` comment should be moved to the <remarks/> section.

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 4 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:142]: Moving to remarks section. Thanks

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 5]
```diff
@@ -583,9 +584,14 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
         /// cannot be created.
         /// </remarks>
         /// <param name="writer">The <see cref="Utf8JsonWriter"/> to write to.</param>
-        /// <param name="value">The value to convert.</param>
+        /// <param name="value">The value to convert. <seealso cref="HandleNull"/> to determine if the converter handles a 'null' reference.</param>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:587]:
```suggestion
        /// <param name="value">The value to convert. <seealso cref="HandleNull"/> to determine if the converter handles <see langword="null" /> values.</param>
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 6]
```diff
@@ -140,6 +140,7 @@ internal virtual bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, J
         /// <param name="typeToConvert">The <see cref="Type"/> being converted.</param>
         /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
         /// <returns>The value that was converted.</returns>
+        /// <remarks><seealso cref="HandleNull"/> to determine if the converter handles a 'null' JSON token.</remarks>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:143]:
```suggestion
        /// <remarks>Note that the value of <seealso cref="HandleNull"/> determines if the converter handles null JSON tokens.</remarks>
```

**Pull request merged**